### PR TITLE
Refactor extension loading to allow for overloads and suppression

### PR DESCRIPTION
**Pull request by Kordyjan**

### CODE REVIEW

File: core/src/main/kotlin/plugability/DokkaContext.kt [hunk 1]
```diff
@@ -86,14 +104,39 @@ private class DokkaContextConfigurationImpl(
...
+                .flatMap { buckets[it].orEmpty() }
+                .toSet()
+            mergedBucket.forEach { buckets[it] = mergedBucket }
         }
+        return buckets.values.distinct().associateBy(::findNotOverridden)
+    }
+
+    private fun findNotOverridden(bucket: Set<Extension<*, *, *>>): Extension<*, *, *> {
+        val filtered = bucket.filter { it !in suppressedExtensions }
+        return filtered.singleOrNull() ?: throw Error("Conflicting overrides: $filtered")
```
Comment (kamildoleglo) [inline core/src/main/kotlin/plugability/DokkaContext.kt:130]: Maybe `IllegalStateException` would be better?

File: core/src/main/kotlin/plugability/DokkaContext.kt [hunk 2]
```diff
@@ -67,14 +67,32 @@ private class DokkaContextConfigurationImpl(
...
+        }
+    }
 
-    private fun topologicalSort() {
+    private val rawExtensions = mutableListOf<Extension<*, *, *>>()
+    private val rawAdjacencyList = mutableMapOf<Extension<*, *, *>, MutableList<Extension<*, *, *>>>()
+    private val suppressedExtensions = mutableMapOf<Extension<*, *, *>, MutableList<Suppression>>()
 
-        val result: MutableList<Extension<*>> = mutableListOf()
+    fun topologicallySortAndPrune() {
```
Comment (kamildoleglo) [inline core/src/main/kotlin/plugability/DokkaContext.kt:84]: I think we could make it `internal`

File: core/src/main/kotlin/plugability/DokkaContext.kt (hunk 1 above)
Comment (Kordyjan) [reply core/src/main/kotlin/plugability/DokkaContext.kt]: done

File: core/src/main/kotlin/plugability/DokkaContext.kt (hunk 2 above)
Comment (Kordyjan) [reply core/src/main/kotlin/plugability/DokkaContext.kt:84]: This function is defined in private class so there is no way to call it from outside of the module.

**Comment by Kordyjan** [comment]
> Please update the developer guide
@MarcinAman done

**Pull request merged**